// utility/array.ts

export function findFirst<T>(arr: Array<T>, comparer: (t: T) => boolean, fallback: T): T {
    var len = arr.length;
    for (var i = 0; i < len; ++i) {
        var v = arr[i];
        if (comparer(v)) return v;
    }
    return fallback;
}

export function sum<T>(arr: Array<T>, fn: (t: T) => number): number {
    return arr.reduce((aggr, d) => aggr + fn(d), 0);
}

export function sum2(arr: Array<number>): number {
    return sum<number>(arr, n => n);
}

export function min<T>(arr: Array<T>, fn: (t: T) => number): T {
    return selectOne<T, number>(arr, fn, (cur: number, min: number) => cur < min);
}

export function max<T>(arr: Array<T>, fn: (t: T) => number): T {
    return selectOne<T, number>(arr, fn, (cur: number, max: number) => cur > max);
}

export function selectOne<T, R>(arr: Array<T>, fn: (t: T) => R, cmp: (cur: R, sele: R) => boolean): T {
    if (arr.length == 0)
        throw new Error('utility:selectOne, bad input, array length could not be zero');
    var mval = fn(arr[0]), mobj = arr[0];
    for (var i = 1; i < arr.length; ++i) {
        var val = fn(arr[i]);
        if (cmp(val, mval)) {
            mobj = arr[i];
            mval = val;
        }
    }
    return mobj;
};

export function avg<T>(nums: Array<T>, mapfn: (t: T) => number): number {
    return nums.length === 0 ? 0 : (sum<T>(nums, mapfn) / nums.length);
}
