// datapvd/factory/raw/growrate.ts

import IFactory from '../fac';
import * as bb from 'bluebird';
import * as def from '../../def';
import * as utility from 'utility';
import * as literal from '../../literal';
import * as facutil from '../facutil';

function dpid(target: string): string { return `GROWRATE_${target}`; }

const growrateFac: IFactory<string, number> = {
    make: (target: string) => {
        return literal.resolve({ type: 'r.end', pack: target })
            .then(epvd => {
                return bb.resolve(new def.StoredDataPvd<number>({
                    id: dpid(target),
                    maxTs: epvd.maxTs,
                    minTs: epvd.remoteTs(epvd.minTs, 1) || facutil.dateTsOffset(epvd.maxTs, 1),
                    hasdef: epvd.hasDef_core,
                    gen: (ts) => {
                        const bts = epvd.remoteTs(ts, -1);
                        return bb.all([epvd.get(ts), epvd.get(bts)]).then(data => {
                            return (<number>data[0] - <number>data[1]) / <number>data[1];
                        })
                    },
                    remoteTs: epvd.remoteTs_core,
                    weakdepts: epvd.weakdepts
                }));
            });
    },
    validate: utility.validate.nonEmptyStr,
    dpid: dpid,
    weakDepts: (target: string) => [target]
};

export default growrateFac;
