// datapvd/factory/raw/grow.ts

import IFactory from '../fac';
import * as bb from 'bluebird';
import * as def from '../../def';
import * as utility from 'utility';
import * as literal from '../../literal';

function dpid(target: string): string { return `GROW_${target}`; }

const growFac: IFactory<string, number> = {
    make: (target: string) => {
        return literal.resolve({ type: 'r.end', pack: target })
            .then(epvd => {
                return bb.resolve(new def.DataPvd<number>({
                    id: dpid(target),
                    maxTs: epvd.maxTs,
                    minTs: epvd.forwardTs(epvd.minTs, 1) || utility.date.dateOffset(epvd.maxTs, { day: 1 }).getTime(),
                    hasdef: (ts: number) => epvd.hasDef(ts),
                    gen: (ts) => {
                        const bts = epvd.backwardTs(ts, 1);
                        return bb.all([epvd.get(ts), epvd.get(bts)]).then(data => {
                            return <number>data[0] - <number>data[1];
                        })
                    }
                }));
            });
    },
    validate: utility.validate.nonEmptyStr,
    dpid: dpid,
    weakDepts: (target: string) => [target]
};

export default growFac;
