// datasrc/wmcloud.ts

import * as bb from 'bluebird';
import * as utility from 'utility';
import * as config from 'config';

interface WMCloudApiRet {
    retCode: number;
    retMsg: string;
    data: Object;
}

export function get<T>(category: string, subcat: string, params?: { [name: string]: string | Array<string> }): bb<T> {
    return bb.resolve()
        .then(() => {
            const parr = new Array<string>();
            params = params || {};
            for (let i in params) {
                const v = params[i];
                if (Array.isArray(v)) {
                    if (v.length > 0) parr.push(`${i}=${v.join(',')}`);
                }
                else {
                    const tv = v.trim();
                    if (tv.length > 0) parr.push(`${i}=${tv}`);
                }
            }
            const pstr = parr.length > 0 ? (`?${parr.join('&')}`) : '';
            return utility.http.webreq({
                method: 'GET',
                host: 'api.wmcloud.com',
                port: 443,
                path: `/data/v1/api/${category}/${subcat}.json${pstr}`,
                headers: {
                    Authorization: `Bearer ${config.wmCloudToken}`
                }
            }, null, true)
                .then(reply => {
                    if (reply.statusCode === 200) throw new Error(`bad status code from wmcloud: ${reply.statusCode}, ${category}, ${subcat}`);
                    const j = <WMCloudApiRet>JSON.parse(reply.data.toString('utf8'));
                    if (j.retCode !== 1) throw new Error(`bad return msg: ${j.retCode}, ${j.retMsg}`);
                    else return <T>j.data;
                });
        });
}

function genFn<T>(category: string, subcat: string) {
    return function (params?: { [name: string]: string | Array<string> }): bb<T> {
        return get<T>(category, subcat, params);
    }
}

function getAllStocks() {
    return master.getSecID({ assetClass: 'E', field: ['listStatusCD', 'secShortName', 'secID', 'listDate', 'exchangeCD'] })
        .then(data => {

        });
}

const market = {
    getMktEqud: genFn('market', 'getMktEqud'),
    getMktIdxd: genFn('market', 'getMktIdxd'),
    //getStockFactorsDateRange: genFn('market', 'getStockFactorsDateRange'),
}
const equity = {
    getEquShare: genFn('equity', 'getEquShare'),
    getEqu: genFn('equity', 'getEqu'),
    getEquIndustry: genFn('equity', 'getEquIndustry'),
};
const idx = {
    getIdx: genFn('idx', 'getIdx'),
};
const master = {
    getSecID: genFn<GetSecIDRet>('master', 'getSecID'),
};

interface GetSecIDRet {
    secID?: string;
    ticker?: string;
    secShortName?: string;
    cnSpell?: string;
    exchangeCD?: string;
    assetClass?: string;
    listStatusCD?: string;
    listDate?: string;
    transCurrCD?: string;
    ISIN?: string;
    partyID?: string;
}
