// datapvd/factory/combine/math.ts

import * as utility from 'utility';
import * as facutil from '../facutil';
import * as def from '../../def';
import * as bb from 'bluebird';

const add = facutil.dpUnion({prefix: 'add', gen: utility.array.sum2});
const sub = facutil.dpUnion({prefix: 'sub', gen: utility.array.sub2});
const mul = facutil.dpUnion({prefix: 'mul', gen: utility.array.mul2});
const div = facutil.dpUnion({prefix: 'div', gen: utility.array.div2});
const max = facutil.dpUnion({prefix: 'max', gen: utility.array.max2});
const min = facutil.dpUnion({prefix: 'min', gen: utility.array.min2});

interface PowPack extends facutil.TransformPackBase<number> {
    exp: number
}
const pow = facutil.dpTransform<PowPack, number, number>({
    prefix: 'POW',
    inputlist: [{ name: 'exp', validate: utility.validate.posNum }],
    gen: (pack: PowPack, dp: def.DataPvd<number>, ts: number, selfdp: def.DataPvd<number>): bb<number> => {
        return dp.get(ts).then(v => Math.pow(v, pack.exp));
    }
})

interface AbsPack extends facutil.TransformPackBase<number> { }
const abs = facutil.dpTransform<AbsPack, number, number>({
    prefix: 'ABS',
    gen: (pack: AbsPack, dp: def.DataPvd<number>, ts: number, selfdp: def.DataPvd<number>): bb<number> => {
        return dp.get(ts).then(Math.abs);
    }
})

export {add, sub, mul, div, max, min, pow, abs};
