// nodes/action/rtplan/simulate.ts


import * as bb from 'bluebird';
import * as utility from 'utility';
import Action from '../action';
import * as log from 'log';
import * as _ from 'lodash';
import * as filestorage from 'filestorage';
import * as dclient from '../../dclient';
import * as mods from 'mods';
import * as nutil from '../../util';
import * as datapvd from 'datapvd';

export interface SimulateInput {
    target: string;
    rtplanId: string;
    redo: boolean;
}

export interface SimulateOutput {
    old: number;
    new: number;
    closed: number;
}

export const action = new Action<SimulateInput, SimulateInput, SimulateOutput>({
    refine: utility.id,
    validate: (input: SimulateInput): boolean => {
        return utility.validate.valueStr(input.target) && utility.validate.valueStr(input.rtplanId) && utility.validate.isBool(input.redo);
    },
    resolve: (input: SimulateInput): bb<SimulateOutput> => {
        return dclient.rtplan.get({ _id: input.rtplanId }, null)
            .then(rtplan => {
                if (!nutil.rtplanScopeContains(input.target, rtplan.targetScope)) throw new Error(`rtplan does not match the target: ${JSON.stringify(rtplan.targetScope)}, ${input.target}`);
                return bb.all([
                    datapvd.literal.resolve({ type: 'r.end', pack: input.target }),
                    datapvd.literal.resolve({ type: 'r.high', pack: input.target }),
                    datapvd.literal.resolve({ type: 'r.low', pack: input.target }),
                    dclient.simtrack.getOrCreate({ target: input.target, rtplanId: input.rtplanId })
                ])
                    .then(data => {
                        const epvd: datapvd.def.DataPvd<number> = data[0], hpvd: datapvd.def.DataPvd<number> = data[1], lpvd: datapvd.def.DataPvd<number> = data[2], simtrack = data[3];
                        if(simtrack.lastSimTs == null || input.redo) {

                        }
                    });
            })
            .then(() => null);
    }
});
