// datapvd/def.ts

import * as co from 'co';
import * as bb from 'bluebird';
import * as utility from 'utility';

export class DataPvd<T> {
    private _id: string;
    private _mints: number;
    private _maxts: number;
    private _hasdef: (ts: number) => boolean;
    private _gen: (ts: number) => bb<T>;

    constructor(pack: { id: string, minTs: number, maxTs: number, hasdef: (ts: number) => boolean, gen: (ts: number) => bb<T> }) {
        if (!utility.validate.nonNegNum(pack.minTs, true) || !utility.validate.nonNegNum(pack.maxTs, true))
            throw new Error(`minTs or maxTs invalid, ${pack.minTs}, ${pack.maxTs}`);
        this._id = pack.id;
        this._maxts = pack.maxTs;
        this._mints = pack.minTs;
        this._hasdef = pack.hasdef;
        this._gen = pack.gen;
    }

    get minTs(): number { return this._mints; }
    get maxTs(): number { return this._maxts; }
    hasDef(ts: number): boolean {
        return ts >= this._mints && ts <= this._maxts && this._hasdef(ts);
    }
    forwardTs(ts: number, n: number): number {
        throw new Error('not implemented');
    }
    backwardTs(ts: number, n: number): number {
        throw new Error('not implemented');
    }
    periodTs(mints: number, maxts: number): Array<number> {
        if (this.hasDef(mints) && this.hasDef(maxts)) {
            var ts = this.forwardTs(mints, 0), ret = new Array<number>();
            maxts = this.backwardTs(maxts, 0);
            if (ts > maxts) return ret;
            while (true) {
                ret.push(ts);
                if (ts === maxts) return ret;
                else ts = this.forwardTs(ts, 1);
            }
        }
        else throw new Error(`input for periodTs is out of defined area, ${mints}, ${maxts}`);
    }
    period(mints: number, maxts: number): bb<Array<{ val: T, ts: number }>> {
        return bb.map(this.periodTs(mints, maxts), ts => this._gen(ts).then(val => { return { ts: ts, val: val }; }));
    }
    get(ts: number): bb<T> {
        if (this.hasDef(ts)) return this._gen(ts);
        else throw new Error(`input for get is out of defined area, ${ts}`);
    }
}
