// datapvd/factory/basic/boll.ts

import IFactory from '../fac';
import * as bb from 'bluebird';
import * as def from '../../def';
import * as utility from 'utility';
import * as literal from '../../literal';
import * as facutil from '../facutil';

function dpid(pack: BollFacPack): string { return `BOLL_${pack.N}_${pack.W}_${literal.dpid(pack.dp)}`; }

interface BollFacPack {
    N: number,
    W: number,
    dp: def.DataPvd<number> | literal.LiteralDP
}

interface BollRet {
    up: number,
    mid: number,
    low: number
}

const bollFac: IFactory<BollFacPack, BollRet> = {
    make: (pack: BollFacPack) => {
        return literal.resolve(pack.dp)
            .then(dp => bb.all([dp, literal.resolve({ type: 'b.ma', pack: { N: pack.N, dp: dp } })]))
            .then(data => {
                const mapvd = <def.DataPvd<number>>data[1], dp = <def.DataPvd<number>>data[0];
                return bb.resolve(new def.StoredDataPvd<BollRet>({
                    id: dpid(pack),
                    maxTs: dp.maxTs,
                    minTs: dp.remoteTs(dp.minTs, 2 * pack.N - 2) || facutil.dateTsOffset(dp.maxTs, 1),
                    hasdef: (ts: number) => dp.hasDef(ts),
                    gen: (ts: number): bb<BollRet> => {
                        const bts = dp.remoteTs(ts, -pack.N + 1);
                        return bb.all([dp.period(bts, ts), mapvd.period(bts, ts)])
                            .then(data => {
                                // (MA[(dp - ma)^2])^(1/2)
                                const W = pack.W;
                                const malist = data[1], dpvallist = data[0];
                                const maval = malist[malist.length - 1].val;
                                const devi = Math.pow(utility.array.avg(dpvallist.map((v, idx) => Math.pow(v.val - malist[idx].val, 2)), t => t), 0.5);
                                return { up: maval + devi * W, mid: maval, low: maval - devi * W };
                            });
                    },
                    remoteTs: dp.remoteTs_core,
                    weakdepts: dp.weakdepts
                }));
            });
    },
    validate: (pack: BollFacPack): boolean => utility.validate.posNum(pack.N, true) && utility.validate.posNum(pack.W, true) && literal.validate(pack.dp),
    dpid: dpid,
    weakDepts: (pack: BollFacPack): Array<string> => literal.weakDepts(pack.dp)
};

export default bollFac;
