// nodes/def.ts

import * as mongo from 'mongodb';

export interface Task {
    _id?: mongo.ObjectID | string;
    locality?: Object;
    condition?: { type: string, pack?: any };
    action?: { type: string, pack?: any };
    postAction?: Object;
    constraints?: { ttl?: number, conditionCheckInterval?: number, timeoutLevel?: number };
    comments?: string;
    statusId?: number;
    createdTs?: number;
    processLog?: Array<Object>;
    lastProcessTs?: number;
    nextConditionCheckTs?: number;
    lastConditionCheckTs?: number;
    quickview?: Object;
    priority?: number;
    assigned?: number;
}

export interface Simulate {
    _id?: mongo.ObjectID | string;
    closed?: boolean;
    target?: string;
    sts?: number;
    sp?: number;
    ets?: number;
    ep?: number;
    hts?: number;
    hp?: number;
    lts?: number;
    lp?: number;
    glong?: number;
    rtplanId?: string;
    env?: Array<Object>;
    concernsIn?: Array<Object>;
    concernsOut?: Array<Object>;
}

export interface Rtprog {
    _id?: mongo.ObjectID | string;
    rtplanId?: string;
    target?: string;
    rtprog?: Object;
    glong?: boolean;
    createdTs: number;
}

export interface RtprogOut {
    _id?: mongo.ObjectID | string;
    rtplanId?: string;
    target?: string;
    rtprog?: Object;
    glong?: boolean;
    createdTs: number;
    sts?: number;
    sp?: number;
    hts?: number;
    hp?: number;
    lts?: number;
    lp?: number;
}
